contracts/Router.sol

```diff
@@ -0,0 +1,240 @@
+    /// @notice the minimum value of charityCut
+    uint256 public constant MIN_CHARITY_CUT = 10**17; // 10%
+    /// @notice the maximum value of charityCut
+    uint256 public constant MAX_CHARITY_CUT = 5 * 10**17; // 50%
+    /// @notice the minimum value of rewardsCut
+    uint256 public constant MIN_REWARDS_CUT = 5 * 10**15; // 0.5%
+    /// @notice the maximum value of rewardsCut
+    uint256 public constant MAX_REWARDS_CUT = 10**17; // 10%
+
+    address private gov;
+    address private charity;
+    uint256 private charityCut;
+    uint256 private rewardsCut;
+
+    I_ERC20 public tree = I_ERC20(TREE);
+    I_ERC20 public reserveToken = I_ERC20(DAI);
+    I_TREERewards public lpRewards;
+
+    uint256 private totalPledged;
+    uint256 private numPledgers;
+    uint256 private treeSold;
+    mapping (uint256 => address) private pledgers;
+    mapping (address => uint256) private amountsPledged;
+ 
+    constructor(address _gov, address _charity, address _lpRewards, uint256 _charityCut, uint256 _rewardsCut) public {
+        gov = _gov;
+        charity = _charity;
+        lpRewards = _lpRewards;
+        charityCut = _charityCut;
+        rewardsCut = _rewardsCut;
+    }
+
+
+    function pledge(uint256 _amount, bool max) external payable {
+        require(!Address.isContract(msg.sender), "Must pledge from EOA");
+        if (max) {_amount = reserveToken.balanceOf(msg.sender);}
+        require(_amount > 0, "Must pledge more than 0.");
+        require(reserveToken.balanceOf(msg.sender) >= _amount, "Cannot pledge more reserveToken than held.");
+        reserveToken.transferFrom(msg.sender, address(this), _amount);
+
+        totalPledged = totalPledged + _amount;
+
+        uint256 pledgerId = getPledgerId(msg.sender);
+        if (pledgerId == 0) {
+            // user has not pledged before
+            pledgerId = numPledgers++;
+            pledgers[pledgerId] = msg.sender;
+        }
+        amountsPledged[msg.sender] = amountsPledged[msg.sender].add(_amount);
+
+        emit Pledge(msg.sender, _amount);
+    }
+
+
+    function unpledge(uint256 _amount, bool max) external payable {
+
+        require(hasPledged(msg.sender), "User has not pledged.");
+        if (max) {_amount = amountsPledged[msg.sender];}
+        require(_amount <= amountsPledged[msg.sender], "Cannot unpledge more than already pledged.");
+
+        totalPledged = totalPledged.sub(_amount);
+        amountsPledged[msg.sender] = amountsPledged.sub(_amount);
+
+        reserveToken.transfer(msg.sender, _amount);
+
+        emit Unpledge(msg.sender, _amount);
+    }
+
+
+    function swapExactTokensForTokens(
+        uint amountIn,
+        uint amountOutMin,
+        address[] calldata path,
+        address to,
+        uint deadline
+    ) external override returns (uint256[] memory amounts) {
+        require(deadline >= block.timestamp, 'UniswapV2Router: EXPIRED');
+
+        require(totalPledged >= amountIn, "Not enough DAI pledged. Rebase postponed.");
```

Comment:
We should restrict calling this function to `RESERVE` only; otherwise anyone could call it with a lower `amountIn` value and trigger a smaller rebase before a rebase should be eligible.